README.md
# blurhash

This library is a [Blurhash](https://github.com/woltapp/blurhash/blob/master/Algorithm.md) algorithm implementation for dart. For more information or other implementations go to https://github.com/woltapp/blurhash.

Instead of showing boring placeholders, just show an idea of the picture while the actual picture is loading! Images can be converted to Base83 on your backend or while uploading in the frontend. Afterwards, you can send the Base83 string to your frontend. The client can then decode the Base83 string and draw the blurred image.

|            | Original                         | Blurred                          |
| ---------- |:------------------------------:| :-----------------------------:|
| **Image**  | <img src="resources/unsplash-image.jpg" alt="original" width="210" height=130/>             |<img src="resources/blurred.jpg" alt="original" width="255" height=130/> |

Based on the hash `q.NK3Mt7WrofayazbHj[l.TkCWBWCj[j@f6azIUWXjZWBWCjsoLayM{ofazayjZa#Wqj[kCWBj[bHWXj[jZWVt7WCs:ofa}axjZay`

##### Example usage
See the example in [example_main](example/lib/example_extended.dart)

<img src="resources/blur.gif" alt="original" width="255"/>

## Getting started
Include the library in your project. Then call the decoder with a valid Base83 string

```Dart
Image.memory(blurhash.Decoder.decode(hash, 300, 200))
```


Credits go to [Dag Ågren](https://github.com/DagAgren) / [Wolt](https://github.com/woltapp) for this idea.

CHANGELOG.md
## [0.1.0] - TODO: Add release date.

* Initial release with a decoder for blur hashes plus an example app


test/base83_test.dart
import 'package:blurhash/blurhash.dart';
import 'package:test/test.dart';

void main() {
  test('Decode base83 empty string', () async {
    int result = Base83.decode('');
    expect(result, 0);
  });

  test('Decode base83 "foobar"', () async {
    int result = Base83.decode('foobar');
    expect(result, 163902429697);
  });

  test('Decode base83 "LFE.@D9F01_2%L%MIVD*9Goe-;WB"', () async {
    int result = Base83.decode(r'LFE.@D9F01_2%L%MIVD*9Goe-;WB');
    expect(result, -1597651267176502418);
  });
}


lib/src/base83.dart
part of blurhash;

/// Implementation for the custom Base83 algorithm. See https://github.com/woltapp/blurhash/blob/master/Algorithm.md#base-83.
class Base83 {
  /// Decode a given string from position to position from a valid Base63.
  static int decode(String str, {int from = 0, int to = -1}) {
    int result = 0;
    if (to == -1) {
      to = str.length;
    }
    for (int i = from; i < to; i++) {
      int index = _charMap[str[i]] ?? -1;
      if (index != -1) {
        result = result * 83 + index;
      }
    }
    return result;
  }

  static const Map<String, int> _charMap = <String, int>{
    '0': 0,
    '1': 1,
    '2': 2,
    '3': 3,
    '4': 4,
    '5': 5,
    '6': 6,
    '7': 7,
    '8': 8,
    '9': 9,
    'A': 10,
    'B': 11,
    'C': 12,
    'D': 13,
    'E': 14,
    'F': 15,
    'G': 16,
    'H': 17,
    'I': 18,
    'J': 19,
    'K': 20,
    'L': 21,
    'M': 22,
    'N': 23,
    'O': 24,
    'P': 25,
    'Q': 26,
    'R': 27,
    'S': 28,
    'T': 29,
    'U': 30,
    'V': 31,
    'W': 32,
    'X': 33,
    'Y': 34,
    'Z': 35,
    'a': 36,
    'b': 37,
    'c': 38,
    'd': 39,
    'e': 40,
    'f': 41,
    'g': 42,
    'h': 43,
    'i': 44,
    'j': 45,
    'k': 46,
    'l': 47,
    'm': 48,
    'n': 49,
    'o': 50,
    'p': 51,
    'q': 52,
    'r': 53,
    's': 54,
    't': 55,
    'u': 56,
    'v': 57,
    'w': 58,
    'x': 59,
    'y': 60,
    'z': 61,
    '#': 62,
    r'$': 63,
    '%': 64,
    '*': 65,
    '+': 66,
    ',': 67,
    '-': 68,
    '.': 69,
    ':': 70,
    ';': 71,
    '=': 72,
    '?': 73,
    '@': 74,
    '[': 75,
    ']': 76,
    '^': 77,
    '_': 78,
    '{': 79,
    '|': 80,
    '}': 81,
    '~': 82,
  };
}


lib/src/decoder.dart
part of blurhash;

/// Encoder class for the blur hash.
class Decoder {
  /// Decode a given Base83 string into a Uint8List representing the image.
  static Uint8List decode(String blurHash, int width, int height,
      {double punch = 1.0}) {
    if (blurHash == null || blurHash.length < 6) {
      return null;
    }
    int numCompEnc = Base83.decode(blurHash, from: 0, to: 1);
    int numCompX = (numCompEnc % 9) + 1;
    int numCompY = ((numCompEnc / 9) + 1).floor();
    if (blurHash.length != 4 + 2 * numCompX * numCompY) {
      return null;
    }
    int maxAcEnc = Base83.decode(blurHash, from: 1, to: 2);
    double maxAc = (maxAcEnc + 1) / 166.0;

    List<List<double>> colors = List<List<double>>((numCompX * numCompY));
    for (int i = 0; i < colors.length; i++) {
      if (i == 0) {
        int colorEnc = Base83.decode(blurHash, from: 2, to: 6);
        colors[i] = _decodeDc(colorEnc);
      } else {
        int from = 4 + i * 2;
        int colorEnc = Base83.decode(blurHash, from: from, to: from + 2);
        colors[i] = _decodeAc(colorEnc, maxAc * punch);
      }
    }
    Uint8List bitmap =
        _composeBitmap(width, height, numCompX, numCompY, colors);
    _RGBA32BitmapHeader header =
        _RGBA32BitmapHeader(bitmap.length, width, height);

    return header.appendContent(bitmap);
  }

  /// Decode a given Base63 string into a dart:ui Image
  static Future<ui.Image> decodeAsImage(String blurHash, int width, int height,
      {double punch = 1.0}) async {
    Uint8List list = decode(blurHash, width, height, punch: punch);

    return ui
        .instantiateImageCodec(list, targetWidth: width, targetHeight: height)
        .then((ui.Codec c) => c.getNextFrame())
        .then((ui.FrameInfo i) => i.image);
  }

  static List<double> _decodeDc(final int colorEnc) {
    final int r = colorEnc >> 16;
    final int g = (colorEnc >> 8) & 255;
    final int b = colorEnc & 255;
    return <double>[_srgbToLinear(r), _srgbToLinear(g), _srgbToLinear(b)];
  }

  static double _srgbToLinear(final int colorEnc) {
    final double v = colorEnc / 255.0;

    if (v <= 0.04045) {
      return (v / 12.92);
    } else {
      return pow((v + 0.055) / 1.055, 2.4).toDouble();
    }
  }

  static List<double> _decodeAc(int value, double maxAc) {
    double r = value / (19 * 19).floor();
    double g = (value / 19) % 19.floor();
    int b = value % 19;
    return <double>[
      _signedPow2((r - 9) / 9.0) * maxAc,
      _signedPow2((g - 9) / 9.0) * maxAc,
      _signedPow2((b - 9) / 9.0) * maxAc
    ];
  }

  static int _linearToSrgb(double value) {
    double v;
    if (value > 1) {
      v = 1.0;
    } else if (value < 0) {
      v = 0.0;
    } else {
      v = value;
    }
    if (v <= 0.0031308) {
      return (v * 12.92 * 255 + 0.5).round();
    } else {
      return ((1.055 * pow(v, 1 / 2.4) - 0.055) * 255 + 0.5).round();
    }
  }

  static double _signedPow2(double value) {
    return value.sign * pow(value.abs(), 2.0);
  }

  static Uint8List _composeBitmap(int width, int height, int numCompX,
      int numCompY, List<List<double>> colors) {
    Int32List list = Int32List(width * height);
    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        double r = 0.0;
        double g = 0.0;
        double b = 0.0;
        for (int j = 0; j < numCompY; j++) {
          for (int i = 0; i < numCompX; i++) {
            double basis =
                (cos(math.pi * x * i / width) * cos(math.pi * y * j / height))
                    .toDouble();
            List<double> color = colors[j * numCompX + i];
            r += color[0] * basis;
            g += color[1] * basis;
            b += color[2] * basis;
          }
        }
        int index = y * width + x;
        ui.Color color = ui.Color.fromRGBO(
            _linearToSrgb(b), _linearToSrgb(g), _linearToSrgb(r), 1.0);
        list[index] = color.value;
      }
    }
    return list.buffer.asUint8List();
  }
}

const int _RGBA32HeaderSize = 122;

class _RGBA32BitmapHeader {
  final int contentSize;
  Uint8List headerIntList;

  /// Create a new RGBA32 header
  _RGBA32BitmapHeader(this.contentSize, int width, int height) {
    final int fileLength = contentSize + _RGBA32HeaderSize;
    headerIntList = Uint8List(fileLength);
    headerIntList.buffer.asByteData()
      ..setUint8(0x0, 0x42)
      ..setUint8(0x1, 0x4d)
      ..setInt32(0x2, fileLength, Endian.little)
      ..setInt32(0xa, _RGBA32HeaderSize, Endian.little)
      ..setUint32(0xe, 108, Endian.little)
      ..setUint32(0x12, width, Endian.little)
      ..setUint32(0x16, -height, Endian.little)
      ..setUint16(0x1a, 1, Endian.little)
      ..setUint32(0x1c, 32, Endian.little)
      ..setUint32(0x1e, 3, Endian.little)
      ..setUint32(0x22, contentSize, Endian.little)
      ..setUint32(0x36, 0x000000ff, Endian.little)
      ..setUint32(0x3a, 0x0000ff00, Endian.little)
      ..setUint32(0x3e, 0x00ff0000, Endian.little)
      ..setUint32(0x42, 0xff000000, Endian.little);
  }

  Uint8List appendContent(Uint8List contentIntList) {
    headerIntList.setRange(
      _RGBA32HeaderSize,
      contentSize + _RGBA32HeaderSize,
      contentIntList,
    );

    return headerIntList;
  }
}


example/lib/example.dart
import 'package:blurhash/blurhash.dart' as blurhash;
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter blurhash',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter blurhash'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // A correct blurhash
  static const hash =
      r"q.NK3Mt7WrofayazbHj[.TkCWBWCj[j@f6azIUWXjZWBWCjsoLayM{ofazayjZa#Wqj[kCWBj[bHWXj[jZWVt7WCs:ofa}axjZay";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Image.memory(blurhash.Decoder.decode(hash, 300, 200)),
      ),
    );
  }
}


example/lib/example_extended.dart
import 'package:blurhash/blurhash.dart' as blurhash;
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter blurhash',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter blurhash'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // A correct blurhash
  static const hash =
      r"q.NK3Mt7WrofayazbHj[.TkCWBWCj[j@f6azIUWXjZWBWCjsoLayM{ofazayjZa#Wqj[kCWBj[bHWXj[jZWVt7WCs:ofa}axjZay";
  bool _showBlur = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.center,
          children: <Widget>[
            GestureDetector(
              onTap: () {
                setState(() {
                  _showBlur = !_showBlur;
                });
              },
              child: Stack(
                alignment: Alignment.center,
                children: <Widget>[
                  Image.asset(
                    "<put a valid file>",
                    width: 300,
                    height: 200,
                  ),
                  AnimatedOpacity(
                    opacity: _showBlur ? 1.0 : 0.0,
                    duration: const Duration(milliseconds: 300),
                    child:
                        Image.memory(blurhash.Decoder.decode(hash, 300, 200)),
                  ),
                ],
              ),
            ),
            SizedBox(
              height: 50,
            ),
            Text(
              "Tap image!",
              style: TextStyle(fontSize: 20),
            )
          ],
        ),
      ),
    );
  }
}
